File: backend/app/serializers.py
from app.extensions import db
from app.models import Club, Event, User


def user_to_dict(u: User) -> dict:
    from app.models import ClubMember
    joined_ids = [m.club_id for m in ClubMember.query.filter_by(user_id=u.id).all()]
    d = {
        "id": u.id,
        "name": u.name,
        "email": u.email,
        "role": u.role,
        "joinedClubIds": joined_ids,
    }
    if u.club_id:
        d["clubId"] = u.club_id
    return d


def club_to_dict(c: Club) -> dict:
    head = db.session.get(User, c.head_id) if c.head_id else None
    return {
        "id": c.id,
        "name": c.name,
        "description": c.description or "",
        "category": c.category or "",
        "memberCount": c.member_count if c.member_count is not None else 0,
        "headId": c.head_id,
        "headName": head.name if head else "",
        "headEmail": head.email if head else "",
        "createdAt": c.created_at,
        **({"logo": c.logo} if c.logo else {}),
    }


def event_to_dict(e: Event) -> dict:
    club = db.session.get(Club, e.club_id)
    club_name = club.name if club else ""
    d = {
        "id": e.id,
        "title": e.title,
        "description": e.description or "",
        "date": e.date,
        "time": e.time,
        "location": e.location,
        "clubId": e.club_id,
        "clubName": club_name,
        "status": e.status,
        "createdBy": e.created_by,
    }
    if e.attendance_count is not None:
        d["attendanceCount"] = e.attendance_count
    return d

File: backend/src/serializers.js
const User = require('./models/User');
const Club = require('./models/Club');
const ClubMember = require('./models/ClubMember');

async function userToDict(u) {
  const memberships = await ClubMember.find({ userId: u._id });
  const joinedClubIds = memberships.map((m) => m.clubId);
  const d = {
    id: u._id,
    name: u.name,
    email: u.email,
    role: u.role,
    joinedClubIds,
  };
  if (u.clubId) {
    d.clubId = u.clubId;
  }
  return d;
}

async function clubToDict(c) {
  const head = c.headId ? await User.findById(c.headId) : null;
  const d = {
    id: c._id,
    name: c.name,
    description: c.description || '',
    category: c.category || '',
    memberCount: c.memberCount != null ? c.memberCount : 0,
    headId: c.headId,
    headName: head ? head.name : '',
    headEmail: head ? head.email : '',
    createdAt: c.createdAt,
  };
  if (c.logo) {
    d.logo = c.logo;
  }
  return d;
}

async function eventToDict(e) {
  const club = await Club.findById(e.clubId);
  const d = {
    id: e._id,
    title: e.title,
    description: e.description || '',
    date: e.date,
    time: e.time,
    location: e.location,
    clubId: e.clubId,
    clubName: club ? club.name : '',
    status: e.status,
    createdBy: e.createdBy,
  };
  if (e.attendanceCount != null) {
    d.attendanceCount = e.attendanceCount;
  }
  return d;
}

module.exports = { userToDict, clubToDict, eventToDict };
